pkg/clients/azure/dns.go

```diff
@@ -0,0 +1,236 @@
+	return "DNS Zone"
+}
+
+func (c *azureClient) AnswerDNSChallenge(reqLogger logr.Logger, acmeChallengeToken string, domain string, cr *certmanv1alpha1.CertificateRequest) (fqdn string, err error) {
+	zone, err := c.zonesClient.Get(context.TODO(), c.resourceGroupName, cr.Spec.ACMEDNSDomain)
+	if err != nil {
+		reqLogger.Error(err, fmt.Sprintf("Error getting dns zone %v", cr.Spec.ACMEDNSDomain))
+		return "", err
+	}
+
+	txtRecordName := c.generateTxtRecordName(domain, *zone.Name)
+	_, err = c.createTxtRecord(reqLogger, txtRecordName, acmeChallengeToken, *zone.Name)
+
+	if err != nil {
+		reqLogger.Error(err, "Error adding acme challenge DNS entry")
+		return "", err
+	}
+
+	reqLogger.Info(fmt.Sprintf("record set added: %v in DNS Zone: %v", txtRecordName, *zone.Name))
+
+	return txtRecordName + "." + cr.Spec.ACMEDNSDomain, nil
+}
+
+func (c *azureClient) DeleteAcmeChallengeResourceRecords(reqLogger logr.Logger, cr *certmanv1alpha1.CertificateRequest) error {
+	zone, err := c.zonesClient.Get(context.TODO(), c.resourceGroupName, cr.Spec.ACMEDNSDomain)
+	if err != nil {
+		reqLogger.Error(err, fmt.Sprintf("Error getting dns zone %v", cr.Spec.ACMEDNSDomain))
+		return err
+	}
+
+	for _, dnsName := range cr.Spec.DnsNames {
+		txtRecordName := c.generateTxtRecordName(dnsName, cr.Spec.ACMEDNSDomain)
+
+		reqLogger.Info(fmt.Sprintf("Deleting record set %v in DNS ZONE: %v", txtRecordName, *zone.Name))
+		_, err = c.recordSetsClient.Delete(context.TODO(), c.resourceGroupName, *zone.Name, txtRecordName, dns.TXT, "")
+
+		if err != nil {
+			reqLogger.Error(err, "Error deleting DNS record: %v from DNS Zone: %v", txtRecordName, *zone.Name)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ValidateDnsWriteAccess spawns a zones client to retrieve the baseDomain's hostedZoneOutput
+// and attempts to write a test TXT ResourceRecord to it. If successful, will return `true, nil`.
+func (c *azureClient) ValidateDNSWriteAccess(reqLogger logr.Logger, cr *certmanv1alpha1.CertificateRequest) (bool, error) {
+
+	zone, err := c.zonesClient.Get(context.TODO(), c.resourceGroupName, cr.Spec.ACMEDNSDomain)
+
+	if err != nil {
+		reqLogger.Error(err, fmt.Sprintf("Error getting dns zone %v", cr.Spec.ACMEDNSDomain))
+		return false, err
+	}
+
+	recordKey := "_certman_access_test." + *zone.Name
+
+	if zone.ZoneType != "Private" {
+		// Build the test record
+		_, err := c.createTxtRecord(reqLogger, recordKey, "\"txt_entry\"", *zone.Name)
+
+		if err != nil {
+			return false, err
+		}
+
+		// After successfull write test clean up the test record and test deletion of that record.
+		_, err = c.recordSetsClient.Delete(context.TODO(), c.resourceGroupName, *zone.Name, recordKey, dns.TXT, "")
+
+		if err != nil {
+			reqLogger.Error(err, "Error while deleting Write Access record")
+			return false, err
+		}
+		// If Write and Delete are successfull return clean.
+		return true, nil
+	}
+
+	reqLogger.Error(err, "Private DNS zone is not allowed")
+	return false, nil
```

Comment:
Personally, I think it makes more sense with this logic reversed. Is this something you considered?

```suggestion
	if zone.ZoneType != "Private" {
		reqLogger.Error(err, "Private DNS zone is not allowed")
		return false, nil
	}
	
	// Build the test record
	_, err := c.createTxtRecord(reqLogger, recordKey, "\"txt_entry\"", *zone.Name)

	if err != nil {
		return false, err
	}

	// After successfull write test clean up the test record and test deletion of that record.
	_, err = c.recordSetsClient.Delete(context.TODO(), c.resourceGroupName, *zone.Name, recordKey, dns.TXT, "")

	if err != nil {
		reqLogger.Error(err, "Error while deleting Write Access record")
		return false, err
	}
	// If Write and Delete are successfull return clean.
	return true, nil
```

Reply:
Agreed!